File: frontend/src/types/events.ts
export const LISTEN_EVENT_TYPES = ['play', 'skip', 'complete'] as const

export type ListenEventType = (typeof LISTEN_EVENT_TYPES)[number]

export interface ListenEventPayload {
  track_id: number
  event_type: ListenEventType
  fraction_played: number
}

export interface TopTrackEntry {
  track: Track
  play_count: number
  listened_seconds: number
}

import type { Track } from './track'

export interface ListeningStats {
  period_days: number
  play_count: number
  skip_count: number
  complete_count: number
  listened_seconds: number
  top_tracks: TopTrackEntry[]
}

export interface ListeningHistoryItem {
  event: {
    id: number
    track_id: number
    event_type: ListenEventType
    fraction_played: number
    created_at: string
  }
  track: Track
}

export interface ListeningHistoryResponse {
  items: ListeningHistoryItem[]
  total: number
  limit: number
  offset: number
}

File: backend/app/schemas/events.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.tracks import TrackResponse


ListenEventType = Literal["play", "skip", "complete"]

COMPLETE_MIN_FRACTION = 0.9


class ListenEventCreate(BaseModel):
    track_id: int = Field(ge=1)
    event_type: ListenEventType
    fraction_played: float = Field(ge=0.0, le=1.0)

    @model_validator(mode="after")
    def validate_complete_fraction(self) -> "ListenEventCreate":
        if self.event_type == "complete" and self.fraction_played < COMPLETE_MIN_FRACTION:
            raise ValueError(
                f"complete events require fraction_played >= {COMPLETE_MIN_FRACTION}"
            )
        return self


class ListenEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    track_id: int
    event_type: ListenEventType
    fraction_played: float
    created_at: datetime


class TopTrackEntry(BaseModel):
    track: TrackResponse
    play_count: int
    listened_seconds: float


class ListeningStatsResponse(BaseModel):
    period_days: int
    play_count: int
    skip_count: int
    complete_count: int
    listened_seconds: float
    top_tracks: list[TopTrackEntry]


class ListeningHistoryItem(BaseModel):
    event: ListenEventResponse
    track: TrackResponse


class ListeningHistoryResponse(BaseModel):
    items: list[ListeningHistoryItem]
    total: int
    limit: int
    offset: int
